Fix part of #19435: Learner group editor page migration

AFZL210 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *all* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes part of #19435 .
2. This PR does the following: Migrates the ```Learner Group Editor``` page from Webpack/AngularJS to Lazy-Loaded Angular Modules

## Essential Checklist

- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [x] I have followed the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- [x] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I don't have permissions to assign reviewers directly).
- [x] The linter/Karma presubmit checks pass on my local machine, and my PR follows the [coding style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide)).
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)


## Proof that changes are correct

when user is a valid facilitator

[Screencast from 16-02-24 08:09:53 PM IST.webm](https://github.com/oppia/oppia/assets/79896602/d13acfc8-8727-4bb4-8120-7afda28d5eb6)


logged out user

[Screencast from 16-02-24 08:25:51 PM IST.webm](https://github.com/oppia/oppia/assets/79896602/0a47d934-89bd-48e0-b996-438fa266600e)

when the user isn't a valid facilitator

[Screencast from 16-02-24 08:27:05 PM IST.webm](https://github.com/oppia/oppia/assets/79896602/8476277e-239b-4f58-857c-a8ef3a5a7863)


<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface. Please also include
videos/screenshots of the developer tools browser console, so that we can be
sure that there are no console errors.

When you make updates to the PR, please update these videos/screenshots as well.
You can drop videos/screenshots from previous versions of the PR.

The above should be done for all PRs, including short ones (e.g. a single-line change).
However, if the changes in your PRs are autogenerated via a script and you cannot
provide proof for the changes then please leave a comment "No proof of changes
needed because {{Reason}}" and remove all the sections below.
-->

#### Proof of changes on desktop with slow/throttled network
N/A
<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone
N/A
<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language
N/A
<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

AFZL210 committed:
Fix part of #19435: Migrate learner group editor page

AFZL210 committed:
complete migration

AFZL210 committed:
remove browser module

AFZL210 committed:
complete migration

AFZL210 committed:
fix linter

AFZL210 committed:
fix linter

AFZL210 committed:
added frontend test

AFZL210 committed:
add test

AFZL210 committed:
fix linter

AFZL210 committed:
fix test

AFZL210 committed:
fix linter

AFZL210 committed:
fix linter

AFZL210 committed:
fix root

AFZL210 committed:
fix linter

AFZL210 committed:
Merge branch 'develop' into fix19435-migrate-learner-group-editor

AFZL210 committed:
update comment

DubeySandeep commented on core/templates/pages/learner-group-pages/edit-group/edit-learner-group-page-auth.guard.spec.ts:
```diff
@@ -0,0 +1,122 @@
+
+class MockAccessValidationBackendApiService {
+  validateAccessToLearnerGroupEditorPage(learnerGroupId: string) {
+    return of().toPromise();
```
Any specific reason for creating a promise via `of` and not directly `new Promise(...`?

AFZL210 replied:
not really both do the same thing, removed it.

seanlip reviewed (approved):
Thanks, LGTM for codeowners.

DubeySandeep commented on webpack.common.config.ts:
```diff
@@ -442,17 +442,6 @@ module.exports = {
       minify: htmlMinifyConfig,
       inject: false
     }),
-    new HtmlWebpackPlugin({
```
Remove `learner_group_editor` from the entry (Around line 96)

AFZL210 replied:
Done.

DubeySandeep commented on core/templates/pages/oppia-root/routing/access-validation-backend-api.service.ts:
```diff
@@ -36,6 +36,11 @@ export class AccessValidationBackendApiService {
     '/access_validation_handler/can_access_release_coordinator_page');
 
+  LEARNER_GROUP_EDITOR_PAGE_ACCESS_VALIDATOR = (
+    // eslint-disable-next-line max-len
+    '/access_validation_handler/can_access_edit_learner_group_page/<learner_group_id>'
```
```suggestion
    '/access_validation_handler/can_access_edit_learner_group_page/' +
    '<learner_group_id>'
```

Generally, avoid using the pragma/bypass flag!

AFZL210 replied:
Done.

DubeySandeep commented on core/templates/pages/oppia-root/routing/app.routing.module.ts:
```diff
@@ -79,6 +79,14 @@ const routes: Route[] = [
     canActivate: [IsLoggedInGuard]
   },
+  {
+    path: AppConstants.
```
Should we use `edit-learner-group` over here and in the module use "learnerGroupId:"? Also, this way we won't have to use contextService (which imports lots of other services) in the auth-guard.

AFZL210 replied:
dropped the use of contextService, Done.

DubeySandeep replied:
Sorry for the missunderstanding but can we do something around:
```
in app.routing do:

{
    path: 'edit-learner-group',
    loadChildren: () => import('....')
    ...
  }
  ```
  
  ```
  In the module config do:
  {
    path: ':learnerGroupId',
    component: ...
  }
  ```
  
  
  Also, we can use `route.paramMap.get('id')` in canActivate to extract id instead of using window. 
  
  WDYT?

AFZL210 replied:
yes sounds good will try it.

AFZL210 replied:
Done

DubeySandeep commented on core/templates/pages/learner-group-pages/edit-group/edit-learner-group-page-auth.guard.ts:
```diff
@@ -0,0 +1,66 @@
+
+/**
+ * @fileoverview Guard that redirects user to 401 error page
+ * if the user is not a valid faccilitator.
```
```suggestion
 * if the user is not a valid facilitator.
```

AFZL210 replied:
Done.

DubeySandeep commented on core/templates/pages/learner-group-pages/edit-group/edit-learner-group-page-auth.guard.ts:
```diff
@@ -0,0 +1,66 @@
+      state: RouterStateSnapshot
+  ): Promise<boolean> {
+    return new Promise<boolean>((resolve) => {
+      let learnerGroupId = this.contextService.getLearnerGroupId();
```
Can we get if from the route instead of context service? (Note: Context service depends on multiple other services, it's better to make auth-guard simple and light.)

AFZL210 replied:
Done.

DubeySandeep commented on assets/constants.ts:
```diff
@@ -7214,6 +7214,24 @@ export default {
       ]
     },
+    "LEARNER_GROUP_EDITOR": {
+      "ROUTE": "edit-learner-group/:learner_group_id",
```
Can you confirm this URL pattern works for both backend and frontend? (In the backend we use `<>` for the params placeholder.)

AFZL210 replied:
yes it do work, its a placeholder anyways. others follow the same pattern

DubeySandeep replied:
Yes, it looks like it's working. But I'm wondering how `<learner_group_id>` and `:learner_group_id` both work in the backend. Do we do any conversions? I'm not sure, do you see any line of code which can explain this? (Just wanted to make sure we know how it's working!)

AFZL210 replied:
@DubeySandeep Yes, it is being handled by url-interpolation.service. It checks for the <string> pattern and replaces it with the provided value ([here](https://github.com/oppia/oppia/blob/19590f45809ee66ab706920c026cffeab7139ab2/core/templates/pages/oppia-root/routing/access-validation-backend-api.service.ts#L120)).

for eg, if we call interpolateUrl with the url ```/userinfo/<username>``` and interpolationValues as username: afzal the function will return /userinfo/afzal

DubeySandeep replied:
I'm not worried about the frontend, but looking at how it's being handled in the backend? All these UR:s are bound to OppiaRoot handler:
https://github.com/oppia/oppia/blob/de8b88488d53ef1c1a8469d4f95fd2c8f1499e09/main.py#L1148-L1160

So how it's reading `:learner_group_id` as placeholder similar to `<learner_group_id>`? Does it support both?

AFZL210 replied:
It did seem to work with ":" as BLOG_AUTHOR_PROFILE_PAGE also uses this, but I didn't find how it is really being handled.

can you help me with this? i guess im missing something.

DubeySandeep replied:
I'll look into it tonight, though this comment should not block the PR, can you create an issue to track this task and we should continue working on getting this PR merged?

AFZL210 replied:
Sure, filled the [issue](https://github.com/oppia/oppia/issues/19776). PTAL, thanks!

DubeySandeep reviewed (changes requested):
@AFZL210 Thanks for the PR, it looks great! I've left a few comments, PTAL!

AFZL210 committed:
removed use of context service and fixed typo

AFZL210 committed:
fix test

AFZL210 committed:
fix test

AFZL210 committed:
fix test

AFZL210 committed:
fix test

AFZL210 committed:
fix test

AFZL210 committed:
remove max-len flag

AFZL210 committed:
use paramMap instead of windowRef

AFZL210 committed:
avoid using hard coded string

AFZL210 commented:
@DubeySandeep addressed all comments, but this [comment](https://github.com/oppia/oppia/pull/19761#discussion_r1493332953) though it is handled 100% as BLOG_AUTHOR_PROFILE_PAGE also has an extra param but I didn't find the exact reason why it's working.

DubeySandeep commented:
@kevintab95 Can you please review this PR?

kevintab95 commented on core/templates/pages/learner-group-pages/edit-group/edit-learner-group-page.mainpage.html:
```diff
@@ -1,38 +0,0 @@
-
-    </oppia-root>
-    @load('pages/footer_js_libs.html')
-    <script>
-      // Use the normal default browser eventlistener without passive
-      // set to false.
-      // TODO(#10749): Remove workaround for passive event listeners when
-      // jQuery supports it.
-      EventTarget.prototype.addEventListener = (
-        EventTarget.prototype.addEventListener._original);
-    </script>
```
Should we add this in the migrated HTML? Like in https://github.com/oppia/oppia/blob/6948ba953caf9ad53662a60a5a9d2c994bdeff16/core/templates/pages/learner-dashboard-page/learner-dashboard-page-root.component.html#L18-L25

AFZL210 replied:
although I don't have much idea about this, looking at the issue and the explanation for this in the previous PR, I believe we should add this.

kevintab95 reviewed (changes requested):
Thanks @AFZL210! Left a comment. PTAL

AFZL210 committed:
add script for passive event

AFZL210 committed:
Merge branch 'develop' into fix19435-migrate-learner-group-editor

DubeySandeep commented:
@AFZL210 Would you like to pick one more issue from #19435 while we work on getting this PR reviewed and merged?

AFZL210 commented:
> @AFZL210 Would you like to pick one more issue from #19435 while we work on getting this PR reviewed and merged?

Sure
